### PR TITLE
Extend String to support wrap/unwrap of header fields

### PR DESCRIPTION
**Pull request by sbeitzel**
## Description
We add an extension on `String` which implements a simplistic `wrap` function to satisfy the definition of wrapping provided in [RFC5322](https://datatracker.ietf.org/doc/html/rfc5322#section-2.2.3). This then gets invoked on the header fields in a message when we're assembling them, in case the process of assembling the fields causes the length to exceed the specified 998 character limit.

## Motivation and Context
This change fixes issue #121.

## How Has This Been Tested?
I have added unit tests for the wrap/unwrap functions.

## Checklist:
- [x] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [x] If applicable, I have updated the documentation accordingly.
- [x] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Sources/SwiftSMTP/String-Wrap.swift [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/// This extension is intended to encapsulate the wrap/unwrap
+/// functionality described at https://datatracker.ietf.org/doc/html/rfc5322#section-2.2.3
+
+extension String {
+
+    /// Given the desired line length, split this string into
+    /// a sequence of lines of nearly that length, then join them
+    /// back together with the sequence `\r\n `
+    public func wrap(_ to: Int = 78) -> String {
```
Comment (dannys42) [inline Sources/SwiftSMTP/String-Wrap.swift:27]: To be consistent with the terminology in RFC5322 2.2.3, we should probably call these functions "foldHeader" and "unfoldHeader".

Also, I think it would be nice to have the parameter be an enum, something like:

 think at this point it would be nice to have a more dedicated API for this purpose.  Perhaps something like:

```swift
enum FoldLineLength {
   /// RFC5322 maximum header length
   case maximum
   /// RFC5322 recommended header length
   case recommended
   case custom(Int)

   var value: Int {
      switch self {
      case .maximum: return 998
      case .recommended: return 78
      case .custom(let length): return length
   }
}
```

The API would then look like this: 

```swift

func foldHeader(_ lineLength: FoldLineLength = .recommended) -> String
```

My intention here is mainly for clarity (i.e. why 78?) and partially to provide for future customizability.

File: Sources/SwiftSMTP/Mail.swift [hunk 2]
```diff
@@ -122,21 +122,21 @@ public struct Mail {
         dictionary["MESSAGE-ID"] = id
         dictionary["DATE"] = Date().smtpFormatted
         dictionary["FROM"] = from.mime
-        dictionary["TO"] = to.map { $0.mime }.joined(separator: ", ")
+        dictionary["TO"] = (to.map { $0.mime }.joined(separator: ", ")).wrap()
```
Comment (dannys42) [inline Sources/SwiftSMTP/Mail.swift:125]: If there was a `wrap()` method on Array, we can replace `joined(separator: ", ").wrap()` with a single `wrap()` call that will be able to ensure comma-separated lists are split at the comma boundaries.  While not required, it will make the header more human readable.

From the RFC:
> if a field body is defined as comma-separated values, it is recommended that folding occur after the comma

File: Sources/SwiftSMTP/String-Wrap.swift (hunk 1 above)
Comment (dannys42) [reply Sources/SwiftSMTP/String-Wrap.swift:27]: In addition, I think these methods should be marked `internal` unless there is any significant need for applications to have access to these methods?

File: Sources/SwiftSMTP/String-Wrap.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            workingString = String(workingString.suffix(from: sliceIndex))
+        }
+        lines.append(workingString)
+        return lines.joined(separator: "\r\n ")
+    }
+
+    /// Find all occurrences of CRLF followed by a whitespace and
+    /// replace them with an empty string.
+    public func unwrap() -> String {
+        let regex = try! NSRegularExpression(pattern: "\r\n\\s", options: [])
```
Comment (dannys42) [inline Sources/SwiftSMTP/String-Wrap.swift:50]: Should `\\s` be `\s`?

File: Tests/SwiftSMTPTests/TestWrapUnwrap.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        for _ in 1...34 {
+            reallyLongLine.append(address)
+            reallyLongLine.append(", ")
+        }
+        reallyLongLine.append(address)
+    }
+
+    func testWrap() throws {
+        let wrapped = reallyLongLine.wrap(78)
+        let lines = wrapped.split(separator: "\r\n")
```
Comment (dannys42) [inline Tests/SwiftSMTPTests/TestWrapUnwrap.swift:35]: For a more complete test, "\r\n" should probably have at least one of the whitespace characters after it.

**Review by dannys42 — COMMENTED** [review]
Thanks for submitting this PR and making a unit test!   Just a few minor comments below.

File: Sources/SwiftSMTP/String-Wrap.swift (hunk 3 above)
Comment (sbeitzel) [reply Sources/SwiftSMTP/String-Wrap.swift:50]: Here the double backslash is to escape the backslash so that it will appear as a single backslash to the regex parser.

File: Tests/SwiftSMTPTests/TestWrapUnwrap.swift (hunk 4 above)
Comment (sbeitzel) [reply Tests/SwiftSMTPTests/TestWrapUnwrap.swift:35]: I'm not entirely sure what you're proposing to test with that change. However, just as a point of clarification, the argument `separator` in the `split` function must be a single character. Swift is interpreting the sequence "\r\n" and a single character, for whatever reason. It sure would be handy if String.split() could act like the split() function in Perl, for example.

File: Sources/SwiftSMTP/String-Wrap.swift (hunk 1 above)
Comment (sbeitzel) [reply Sources/SwiftSMTP/String-Wrap.swift:27]: That's an interesting question. I came to this project because I was looking for a way to build a play-by-email game server entirely in Swift. This project provides half the capability - sending email - but the other half, receiving email, doesn't have a Swift implementation that I could find. A POP3 client *may* have need for such a feature; I honestly have not yet started work on creating such a client, so I don't know.

Comment (dannys42) [reply Sources/SwiftSMTP/String-Wrap.swift:27]: I think it's safe to keep it internal until such a need arises.  I don't like to pollute the namespace especially for common Classes.  It can be jarring for people to see these things as side effects when they're simply importing a library.

File: Tests/SwiftSMTPTests/TestWrapUnwrap.swift (hunk 4 above)
Comment (sbeitzel) [reply Tests/SwiftSMTPTests/TestWrapUnwrap.swift:35]: Actually, I've been thinking further about this and I think that the implementation I've proposed is really inadequate. I will be creating a few more unit tests to demonstrate the proper (RFC) behavior and then see about getting them to pass.

File: Sources/SwiftSMTP/String-Wrap.swift (hunk 1 above)
Comment (sbeitzel) [reply Sources/SwiftSMTP/String-Wrap.swift:27]: No, you're absolutely right, we don't want to make any ugly surprises. Honestly, I'm starting to wonder if an extension on String is even the right way to go, or if it would be better to make a separate function that accepts a String and a desired length. Something like:

```
/// Error type thrown by fold/unfold functions
enum FoldError: Error { case maxLengthExceeded }

/// Fold a header, if necessary, to respect the provided line length
///
/// If the folded header still contains a line that is longer than the
/// maximum length specified, then throw an error.
func fold(_ header: String, at lineLength: FoldLineLength = .recommended,
      maxLength: FoldLineLength = .maximum) throws -> String

// Unfold a header, converting the fold token to a single whitespace.
func unfold(_ header: String) -> String
```

Comment (dannys42) [reply Sources/SwiftSMTP/String-Wrap.swift:27]: I don't think it should be a global function.  As an internal extension on String I think it's fine as long as it's clear that it's for the header... either by naming it `foldHeader()` or `fold(header: String,...)`.  Alternatively we could perhaps have a HeaderValue class/struct that has a "values" array as well as a string conversion for use in a header.  It feels a bit heavy for what we're doing, but it might be a cleaner approach as it would fully encapsulate the different ways in which we can use the header values while still using the dictionary.  It still feels a little heavy for what we're doing though.

Comment (sbeitzel) [reply Sources/SwiftSMTP/String-Wrap.swift]: I've moved the functionality inside Mail and kept it private, so I think that addresses the whole namespace/heaviness problem.

File: Sources/SwiftSMTP/Mail.swift [hunk 5]
```diff
@@ -122,27 +123,72 @@ public struct Mail {
...
+        }
+        if register.count > 997 {
+            Log.error("Header line length exceeds the specified maximum (998 chars) - see RFC 5322 section 2.2.3")
+        }
+        foldedHeader.append(register)
+
+        let valueIndex = foldedHeader.index(foldedHeader.startIndex, offsetBy: key.count + 2)
+        return String(foldedHeader.suffix(from: valueIndex))
+    }
+
```
Comment (dannys42) [inline Sources/SwiftSMTP/Mail.swift:191]: I feel like there's a lot of complexity here that's a little hard to follow.  While your approach is certainly memory efficient, I think it would be good to at separates the two logical concepts: splitting the input value into elements and joining them into proper length lines.

Perhaps it might make things easier to work with arrays?  So, split the input value into elements, trim leading and training whitespaces in each element, then use that to construct a line of text until it reaches the maximum length, and appending that to an array of output lines.  On return, simply join the array with commas.  I see two additional points of complexity remaining:

- An element itself may exceeds the 997 limit.  
- Separating the text into elements.  Section 3.2.2 talks about "folding white space and comments" that I think make this a bit more complicated and probably deserves it's own function.

The first one seems easier to solve as the note in RFC 2.2.3 indicates, it sounds like we can just force a fold anywhere we need, and that should be better than allowing it to go above the maximum.

The second one seems like you need a simple lexer... which I always find annoying to write, but can probably be done with a simple state machine.

Also, I suggest having some `let` constants somewhere representing `recommendedLineLength=78` and `maximumLineLength=997` rather than leaving them as magic numbers.

File: Sources/SwiftSMTP/Mail.swift [hunk 6]
```diff
@@ -122,27 +123,72 @@ public struct Mail {
...
     }
 
+    private func foldHeaderValue(key: String, value: String) -> String {
+        let initialHeader = "\(key): \(value)"
+        if initialHeader.count < 79 {
+            return value
+        }
+        // if we're here, it means that RFC 5322 SUGGESTS that we fold this header
+        var foldedHeader = ""
+        var register = "\(key): "
```
Comment (dannys42) [inline Sources/SwiftSMTP/Mail.swift:154]: Won't this result in the key field showing up in the returned string?  I don't believe that's the original behavior.

File: Tests/SwiftSMTPTests/TestHeaderFolding.swift [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        for header in allHeaders {
+            if foundCC {
+                // we are now at the first line *after* the CC: line
+                XCTAssertEqual(header, " <unfortunate_joe_1234567890_123456789_123456783_123456784_123456785_123456786@dumbster.local>")
+                break
+            } else  if header.hasPrefix("CC:") {
+                foundCC = true
+                XCTAssertEqual(header, "CC: \("Unfortunate Joe".mimeEncoded!) ")
+                continue
+            }
```
Comment (dannys42) [inline Tests/SwiftSMTPTests/TestHeaderFolding.swift:101]: I think this misses the potential case where there is no header after a "CC".

**Review by dannys42 — COMMENTED** [review]
Thanks for iterating and putting thought into this.

File: Sources/SwiftSMTP/Mail.swift [hunk 8]
```diff
@@ -122,27 +123,72 @@ public struct Mail {
         dictionary["MESSAGE-ID"] = id
         dictionary["DATE"] = Date().smtpFormatted
         dictionary["FROM"] = from.mime
-        dictionary["TO"] = to.map { $0.mime }.joined(separator: ", ")
+        dictionary["TO"] = foldHeaderValue(key: "TO", value: to.map { $0.mime }.joined(separator: ", "))
 
         if !cc.isEmpty {
-            dictionary["CC"] = cc.map { $0.mime }.joined(separator: ", ")
+            dictionary["CC"] = foldHeaderValue(key: "CC", value: cc.map { $0.mime }.joined(separator: ", "))
```
[the remark this answers is not in the document]
Comment (sbeitzel) [reply Sources/SwiftSMTP/Mail.swift:129]: I agree. The fact, though, is that the first line of the header has to include the key for length calculation. Moreover, I added logging to the folding function so that if someone adds a header that exceeds the maximum length and thus generates a message that an SMTP server would reject, they'd at least have some clue as to which header is at fault. If I were doing this from scratch, I'd probably rewrite the headers not as a dictionary of name: value but as an array of field structs, where each field has a name property and a value property, and the folding logic is encapsulated inside the field.

To my mind, that seems a bit drastic just to address what is ultimately an edge case.

File: Sources/SwiftSMTP/Mail.swift (hunk 5 above)
Comment (sbeitzel) [reply Sources/SwiftSMTP/Mail.swift:191]: I don't entirely disagree with you, and that's really the first approach I thought of. However as I went down that path, it turned out to generate *more* complexity. This way, at least, there's minimal backtracking and removing and then restoring characters. If you think a state machine and a tokenizer are more readable than a stream processor, I invite you to write them. :)

As for the magic numbers, you're quite right.

File: Sources/SwiftSMTP/Mail.swift (hunk 6 above)
Comment (sbeitzel) [reply Sources/SwiftSMTP/Mail.swift:154]: Except that at the end, we remove the key the colon and the whitespace from the beginning of the string. As I noted above, the point here is that the first line has to include the key and the delimiter in the character count. Doing it this way was easier for me to understand, although I suppose one might be clever with arithmetic and advance the linePosition, but we also want to be sure that we don't wind up in a situation like:

```
HEADER: CRLF
 VALUE
```

File: Tests/SwiftSMTPTests/TestHeaderFolding.swift (hunk 7 above)
Comment (sbeitzel) [reply Tests/SwiftSMTPTests/TestHeaderFolding.swift:101]: The test message has, in addition to FROM, TO, CC, and SUBJECT, two additional headers, X-OBNOXIOUS and X-SHORT. Are you suggesting the test be rewritten to account for the case where the CC header appears last in the list? I guess that's reasonable.

Comment (dannys42) [reply Tests/SwiftSMTPTests/TestHeaderFolding.swift]: Yeah, I'm not sure it'd require a rewrite.  It might just need an extra flag to ensure that the foundCC block got executed.  It's not "pretty" but would be sufficient.

File: Sources/SwiftSMTP/Mail.swift (hunk 5 above)
Comment (dannys42) [reply Sources/SwiftSMTP/Mail.swift]: Yeah, I'm not 100% it'd be more readable either... Since you tried that first, I think I'll just trust that you picked the better approach. :-)  I doubt we'll have a "clean" solution to the problem until someone makes a good parser generator for Swift.

File: Sources/SwiftSMTP/Mail.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (dannys42) [reply Sources/SwiftSMTP/Mail.swift:129]: I agree, that's probably the better approach.  Also agree it's a big change for what we're trying to fix.

File: Tests/SwiftSMTPTests/TestHeaderFolding.swift [hunk 9]
```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright Stephen Beitzel 2021
+ *
```
Comment (dannys42) [inline Tests/SwiftSMTPTests/TestHeaderFolding.swift:3]: On a non-technical note, I need to look into what the norm here for open source projects regarding copyrights of contributions.  I thought it was typical for copyright to belong to the project.  But Kitura's a bit of a different case since a lot of the existing files are copyright by IBM, so I understand following that pattern.  Do you have any experience or strong feelings about this?

File: Sources/SwiftSMTP/Mail.swift (hunk 5 above)
Comment (dannys42) [reply Sources/SwiftSMTP/Mail.swift]: Thinking about this more... I believe the RFC talks about trying not to fold during quotes.  So for example, if we have a "To" line consisting of something like:

```
To: "Doe, Jane <jane.doe@nowhere.com>", "Doe, John <john.doe@nowhere.com>"  
```

We'd ideally want the break outside of the quotes.  Do you see an easy way of taking this into account?

Comment (sbeitzel) [reply Sources/SwiftSMTP/Mail.swift]: For starters, I agree with your understanding of the RFC -- but then as I thought about it, it struck me that we're not just folding the email address lines -- we should be folding any header line that gets too long. Now, in an email address, we know that the quote character should always be matched by another quote character, and these are delimiters for a string that we should treat like an atom. But in some other header it might not be special in that way. Suddenly, it's not just a simple arithmetic problem, it's bigger: the folding code needs to know about the semantics of each header field. Okay, we could do that -- at least we're passing in the header name, so the folding code could look up what the delimiter characters are -- but that gets us back to really wanting a custom state machine for parsing the field.

One approach that occurred to me was that we could say, "If you want folding to work on your field the way it works for a list of email addresses, then let us do it; if you want some other rules to apply, then do it yourself." That seems kind of reasonable, and would probably be acceptable for most cases. However, again, we'd probably want to change how we store headers -- as some smarter object rather than just a String. Also, this starts to feel like a behavior/API change that would go into a major release, rather than a bug fix point release.

And at about this point, I reminded myself that the point of this change is to allow for a reasonably long list of recipients. Is there a place where people are talking about the roadmap? If so, I think this is an excellent candidate for discussion there.

File: Tests/SwiftSMTPTests/TestHeaderFolding.swift (hunk 9 above)
Comment (sbeitzel) [reply Tests/SwiftSMTPTests/TestHeaderFolding.swift:3]: It should be whatever the project community agrees on. For obvious legacy reasons, most of the code has a copyright header for IBM, but I understand it's now been spun out from IBM to exist on its own. As part of that process, did anyone set up an entity for "The Kitura Project"?

My own feeling is, I'm happy to assign copyright to Kitura, or to hold it myself but license it under the Apache license since that's the project's license, but it feels weird to assign it to IBM since they've said they're done sponsoring the project.

File: Sources/SwiftSMTP/Mail.swift (hunk 5 above)
Comment (sbeitzel) [reply Sources/SwiftSMTP/Mail.swift]: Oh, and something I discovered while debugging the code and writing the unit tests: because we MIME encode the name part of the email address, and the way email addresses are written, this will only wind up folding between the name and the address, never inside the name. So, even if we do fold inside a single address, we won't be introducing any significant whitespace.

`"Some Fellow" <fellow@place.local>` - the folding code winds up seeing `Some__Fellow` and would not add extra whitespace into the name, which an email client might pick up when creating a contact card.

Comment (dannys42) [reply Sources/SwiftSMTP/Mail.swift]: There's the slack group... You can try there and it might be good for us to see others thoughts on it.

File: Tests/SwiftSMTPTests/TestHeaderFolding.swift (hunk 9 above)
Comment (dannys42) [reply Tests/SwiftSMTPTests/TestHeaderFolding.swift]: I've been updating existing IBM files to say "Copyright: IBM and Kitura project authors".  So if you want to assign it to Kitura or The Kitura Project, AFAIK this is sufficient.  I'll look around and see what other open source projects are doing.  I was hoping to avoid starting a non-profit, but if we need a more definitive legal existence, I can look into that.  Either way I agree it wouldn't make sense to assign it to IBM at this point.

File: Sources/SwiftSMTP/Mail.swift (hunk 5 above)
Comment (dannys42) [reply Sources/SwiftSMTP/Mail.swift]: In particular I think it would be nice to get this reviewed by other users of Swift-SMTP.

**Comment by dannys42** [comment]
@sbeitzel I had to make some changes to CI, would you mind rebasing?

**Comment by sbeitzel** [comment]
@dannys42 When you say, "rebasing," um, what is it that you would like me to do? I have never performed that git operation before. I've just read some documentation that sort of explains what rebasing does to the history, but I'm not entirely sure what that means in the context of a pull request. Does it mean, create a new branch and merge this PR branch into it via rebase, then close this PR and open a new PR with the new branch? Or something else? I mean, I'm happy to cooperate and make your life easier; I just need to understand what I'm supposed to do.

**Comment by dannys42** [comment]
@sbeitzel Ah sorry... Yeah there's different approaches to this.  My preference is rebasing.  So my strategy here would be to:

```
git pull --rebase origin master
```

This will pull in any changes that have occurred on master.  Then resolve any conflicts that occur (there shouldn't be any in your case).

Then when you're done, you'll have to do a `git push --force` or `git push -f`.  This is because the history of your branch no longer matches your original commit.

Note you normally wouldn't have to do this if there are no merge conflicts (as is the case right now)... but in this case we have to bring in the changes to the CI configuration for the build to run correctly.
